test/icat/test_session_handling.py

```diff
@@ -0,0 +1,116 @@
+from datetime import datetime
+
+from icat.client import Client
+import pytest
+
+from datagateway_api.common.config import config
+from datagateway_api.common.icat.filters import PythonICATWhereFilter
+
+
+class TestSessionHandling:
+    def test_session_id_decorator(self):
+        pass
+
+    def test_get_valid_session_details(
+        self, flask_test_app_icat, valid_icat_credentials_header,
+    ):
+        session_details = flask_test_app_icat.get(
+            "/sessions", headers=valid_icat_credentials_header,
+        )
+
+        # Check username is correct
+        assert (
+            session_details.json["USERNAME"] == f"{config.get_test_mechanism()}/"
+            f"{config.get_test_user_credentials()['username']}"
+        )
+        # Check session ID matches the header from the request
+        assert (
+            session_details.json["ID"]
+            == valid_icat_credentials_header["Authorization"].split()[1]
+        )
+
+        session_expiry_datetime = datetime.strptime(
+            session_details.json["EXPIREDATETIME"], "%Y-%m-%d %H:%M:%S.%f",
+        )
+        current_datetime = datetime.now()
+        time_diff = abs(session_expiry_datetime - current_datetime)
+        time_diff_minutes = time_diff.seconds / 60
+
+        # Allows a bit of leeway for slow test execution
+        assert time_diff_minutes < 120 and time_diff_minutes >= 118
+
+    def test_get_invalid_session_details(
+        self, bad_credentials_header, flask_test_app_icat,
+    ):
+        session_details = flask_test_app_icat.get(
+            "/sessions", headers=bad_credentials_header,
+        )
+
+        assert session_details.status_code == 403
+
+    def test_refresh_session(self, valid_icat_credentials_header, flask_test_app_icat):
+        pre_refresh_session_details = flask_test_app_icat.get(
+            "/sessions", headers=valid_icat_credentials_header,
+        )
```

Comment:
Like the comment regarding the final instrument ID, I agree this is a good idea. But because that GET call can happen midway through a test (e.g. `post_refresh_session_details`), putting the code in a pytest fixture isn't suitable. It _can_ be done that way, but that would mean directly calling a pytest fixture, something I've tried to avoid all through these tests since the purpose of the fixtures is that you shouldn't be directly importing them, pytest should pick them up from a function's signature (or from a `usefixtures()` decorator). The alternative is just to use a ordinary function, but since the code uses fixtures (e.g. `flask_test_app_icat`, `valid_icat_credentials_header`) this isn't possible because it can't pick them up. Again, I could directly import them, but that seems against what pytest is about. Slightly annoying, but considering that GET call is only done 4 times, so not too bad